File: day12/day12.cpp
#include <algorithm>
#include <cctype>
#include <fstream>
#include <vector>
#include <unordered_map>
#include <fmt/format.h>

using namespace std;

using Map = unordered_map<string, vector<string>>;

istream& operator>>(istream& in, Map& m)
{
	m.clear();
	string src, dst;
	while (getline(in, src, '-') && getline(in, dst))
	{
		m[src].push_back(dst);
		m[dst].push_back(src);
	}
	return in;
}

static size_t backtrack(vector<string>& lst, bool assigned, const Map& m)
{
	if (lst.back() == "end")
	{
		return 1;
	}

	size_t count = 0;
	for (const auto& can : m.at(lst.back()))
	{
		bool should_skip = false;
		if (can[0] != toupper(can[0]))
		{
			should_skip = find(lst.cbegin(), lst.cend(), can) != lst.cend();
		}

		if ((should_skip && assigned) || can == "start")
		{
			continue;
		}

		lst.push_back(can);
		count += backtrack(lst, assigned || should_skip, m);
		lst.pop_back();
	}
	return count;
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		fmt::print(stderr, "Usage: {} <filename>\n", argv[0]);
		return 1;
	}

	ifstream in(argv[1]);
	if (!in)
	{
		fmt::print(stderr, "Cannot open {}\n", argv[1]);
		return 1;
	}

	Map m;
	in >> m;
	in.close();

	vector<string> lst;
	lst.push_back("start");
	fmt::print("Part1: {}\n", backtrack(lst, true, m));
	fmt::print("Part2: {}\n", backtrack(lst, false, m));
	return 0;
}

File: day12/day12.py
#!/usr/bin/env python3

def backtrack(cur, lst, assigned, m):
    if cur == "end":
        return 1

    lst.append(cur)
    count = 0
    for i in m.get(cur, []):
        should_skip = i.upper() != i and i in lst
        if should_skip and assigned or i == "start":
            continue

        count += backtrack(i, lst, assigned or should_skip, m)

    lst.pop()

    return count

import sys

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            txt = f.read()

        m = {}
        for line in txt.splitlines():
            src, dst = line.split("-")
            m[src] = m.get(src, []) + [dst]
            m[dst] = m.get(dst, []) + [src]

        print("Part1:", backtrack("start", [], True, m))
        print("Part2:", backtrack("start", [], False, m))
    except FileNotFoundError:
        print("Cannot open {}".format(sys.argv[1]), file=sys.stderr)
        sys.exit(1)
